feat(ring_theory/euclidean_domain): generalize lemmas to PIDs

Ruben-VandeVelde opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Ruben-VandeVelde committed:
feat(algebra/euclidean_domain): define a gcd_monoid instance

Ruben-VandeVelde committed:
feat(ring_theory/euclidean_domain): generalize lemmas to PIDs

Ruben-VandeVelde commented on src/ring_theory/principal_ideal_domain.lean:
```diff
@@ -275,3 +276,57 @@ lemma is_principal_ideal_ring.of_surjective [is_principal_ideal_ring R]
+theorem gcd_is_unit_iff {x y : α} : is_unit (gcd x y) ↔ is_coprime x y :=
+by rw [is_coprime, ←mem_span_pair, ←span_gcd, ←span_singleton_eq_top, eq_top_iff_one]
+
+-- this should be proved for UFDs surely?
```
If someone knows the proof in a UFD, please let me know

Vierkantor commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -101,6 +101,22 @@ lemma mem_span_insert {s : set α} {x y} :
 lemma mem_span_singleton' {x y : α} :
   x ∈ span ({y} : set α) ↔ ∃ a, a * y = x := submodule.mem_span_singleton
 
+lemma span_insert (x) (s : set α) :
```
Could you prove this for submodules in general and then specialize this to ideals?

Vierkantor commented on src/ring_theory/principal_ideal_domain.lean:
```diff
@@ -4,6 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Chris Hughes, Morenikeji Neri
 -/
 import ring_theory.unique_factorization_domain
+import ring_theory.coprime.basic
```
My hobby: sorting the imports.
```suggestion
import ring_theory.coprime.basic
import ring_theory.unique_factorization_domain
```

Vierkantor commented on src/ring_theory/euclidean_domain.lean:
```diff
@@ -3,69 +3,35 @@ Copyright (c) 2018 Mario Carneiro. All rights reserved.
-λ ⟨a, b, h⟩, is_unit_iff_dvd_one.2 $ h ▸ dvd_add ((gcd_dvd_left x y).mul_left _)
-  ((gcd_dvd_right x y).mul_left _)⟩
+@[priority 100]  -- see Note [lower instance priority]
+instance gcd_monoid : gcd_monoid α :=
```
I suspect this will actually cause some issues down the line, if we have a euclidean domain with its own `normalized_gcd_monoid`, in particular one such that `euclidean_domain.gcd` is not normalized. We need to make sure that `@normalized_gcd_monoid.gcd R some_proof_that_this_particular_euclidean_domain_is_n_gcd_m` will unify with `@gcd_monoid.gcd R (euclidean_domain.gcd_monoid R)`, otherwise we can be quite sure that we end up misstating some lemma or usages will fail for initially mysterious reasons.

Unfortunately, that will be quite a bit of work: probably the easiest is to rename the current `euclidean_domain` to something like `euclidean_domain_core` and immediately underneath define a new `euclidean_domain` class extending `gcd_monoid` and `euclidean_domain_core`, turning this instance into a `@[reducible] def euclidean_domain.of_core` (see also the reducible non-instances library note). Then we can restate everything currently using `euclidean_domain.gcd` in terms of `gcd_monoid.gcd`. We're not completely done at this point, since we still require the users to carefully set `normalized_gcd_monoid` instances on euclidean domains so that the `gcd`s coincide. So to be completely neat, we also need to complete the square and add a `normalized_euclidean_domain` extending both (the new) `euclidean_domain` and `normalized_gcd_monoid`.

Ruben-VandeVelde replied:
Started on this - could you take a look to see if it makes sense so far?

Vierkantor replied:
That looks exactly correct, thank you very much!

Vierkantor reviewed (commented):
Nice work! I was just about to do some work along these lines, so thank you for being quicker than me. One big issue unfortunately: the `gcd_monoid` instance as it is currently defined will probably cause some non-defeq diamonds, so we need to fix that first.

Ruben-VandeVelde committed:
wip

Vierkantor commented on src/algebra/euclidean_domain.lean:
```diff
@@ -403,7 +408,29 @@ end lcm
+  lcm_zero_right := lcm_zero_right }
+
+@[reducible] def euclidean_domain.of_core : euclidean_domain β :=
+{ ..(infer_instance : euclidean_domain_core β ),
```
For unfolding reasons, it's better to use "French" quotes to refer to a hypothesis than `infer_instance`.
```suggestion
{ ..‹euclidean_domain_core β›,
```

Alternatively, `(by apply_instance : euclidean_domain_core β)` would also work. `infer_instance` will stick around in the term, while the french quotes or tactic calls will be replaced with their result by the elaborator.

Ruben-VandeVelde committed:
chore(algebra/big_operators): avoid 'abel' tactic

abel pulls in the rationals through norm_num, which is quite heavy for goals that can be solved by simp.

Vierkantor committed:
Fix

Ruben-VandeVelde committed:
chore(category_theory/filtered): speed up some proofs

Vierkantor committed:
Fix

Ruben-VandeVelde committed:
ac_refl

kim-em commented on src/algebra/associated.lean:
```diff
@@ -3,10 +3,11 @@ Copyright (c) 2018 Johannes Hölzl. All rights reserved.
 Authors: Johannes Hölzl, Jens Wagemaker
 -/
-import algebra.big_operators.basic
+--import algebra.big_operators.basic
```
```suggestion
```

kim-em commented on src/algebra/associated.lean:
```diff
@@ -103,9 +104,10 @@ lemma exists_mem_multiset_map_dvd {s : multiset β} {f : β → α} :
+-- appears to be unused, but I can find another place for it.
+--lemma exists_mem_finset_dvd {s : finset β} {f : β → α} :
+--  p ∣ s.prod f → ∃ i ∈ s, p ∣ f i :=
+--hp.exists_mem_multiset_map_dvd
```
What is happening here?

kim-em committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

kim-em committed:
merge master, to pull in sym2 speedups

Ruben-VandeVelde committed:
import

Ruben-VandeVelde committed:
Merge branch 'euclidean-gcd-monoid' into pid-gcd

Ruben-VandeVelde committed:
Update associated.lean

Ruben-VandeVelde committed:
Update associated.lean

kim-em committed:
fix timeout

kim-em committed:
oops

Ruben-VandeVelde committed:
simplicial_cosimplicial_augmented_equiv

Ruben-VandeVelde committed:
sum_first_loses_iff_eq

kim-em committed:
attempting to merge master

kim-em committed:
nope, got the merge wrong

Vierkantor committed:
WIP: unbundle `euclidean_domain` into a mixin class

Vierkantor committed:
chore(group_theory/congruence): lower priority for `con.quotient.decidable_eq`

I was debugging some slow typeclass searches, and it turns out (`add_`)`con.quotient.decidable_eq` wants to be applied to all quotient types, causing a search for a `has_mul` instance before the elaborator can try unifying the `con.to_setoid` relation with the quotient type's relation, so e.g. `decidable_eq (multiset α)` first tries to find a `field (list α)` instance before even trying `multiset.decidable_eq`.

Another approach would be to make `multiset` and/or `con.quotient` irreducible, but that would require a lot more work to ensure we don't break the irreducibility barrier.

Vierkantor committed:
Merge branch 'lower-con-quotient-priority' into pid-gcd

Vierkantor committed:
Fix (still a bit hacky!)

Vierkantor committed:
Don't need the `sorry`s anymore

Vierkantor committed:
Lint fix

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into pid-gcd

Ruben-VandeVelde committed:
???

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
try to speed up cech_nerve

Ruben-VandeVelde committed:
fixup! try to speed up cech_nerve

Ruben-VandeVelde committed:
fixup! try to speed up cech_nerve

Ruben-VandeVelde committed:
fixup! try to speed up cech_nerve

adamtopaz commented on src/algebraic_topology/cech_nerve.lean:
```diff
@@ -61,23 +61,32 @@ namespace simplicial_object
 variables [∀ (n : ℕ) (f : arrow C),
   has_wide_pullback f.right (λ i : ulift (fin (n+1)), f.left) (λ i, f.hom)]
 
+def cech_nerve_map {f g : arrow C} (F : f ⟶ g) : (f.cech_nerve ⟶ g.cech_nerve) :=
```
I haven't checked, but I *guess* that adding a `@[simps]` to this definition would solve the error in the proof below.

Vierkantor replied:
Thanks, that seems to help indeed!

Ruben-VandeVelde committed:
fixup! try to speed up cech_nerve

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into pid-gcd

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into pid-gcd

Vierkantor committed:
`@[simps]` a few definitions to help `obviously` figure out the proof faster

Ruben-VandeVelde committed:
is_sheaf_for_is_sheaf_for'

Ruben-VandeVelde committed:
pullback_obj_obj_of_image_open

Ruben-VandeVelde committed:
on_presheaf

Ruben-VandeVelde committed:
Sheaf_spaces_equivelence_sheaf_sites simps

Ruben-VandeVelde committed:
wip

kim-em commented on src/algebraic_geometry/presheafed_space/has_colimits.lean:
```diff
@@ -94,16 +94,32 @@ def pushforward_diagram_to_colimit (F : J ⥤ PresheafedSpace C) :
   map_id' := λ j,
   begin
+    sorry,
     apply (op_equiv _ _).injective,
```
Expanding out all these proofs is not the solution to whatever the problem is here. :-(

Ruben-VandeVelde replied:
Yeah, it doesn't seem to be going anywhere. I don't have a clue where to start looking for a better fix, though.

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into pid-gcd

Ruben-VandeVelde committed:
fixup! wip
